python/modules/IcePy/ObjectAdapter.cpp

```diff
@@ -552,40 +563,48 @@ adapterWaitForDeactivate(ObjectAdapterObject* self, PyObject* args)
     //
     if(PyThread_get_thread_ident() == _mainThreadId)
     {
-        IceUtil::Monitor<IceUtil::Mutex>::Lock sync(*self->deactivateMonitor);
-
         if(!self->deactivated)
         {
-            if(self->deactivateThread == 0)
+            if(self->deactivateFuture == nullptr)
             {
-                AdapterInvokeThreadPtr t = new AdapterInvokeThread(*self->adapter,
-                                                                   &Ice::ObjectAdapter::waitForDeactivate,
-                                                                   *self->deactivateMonitor, self->deactivated);
-                self->deactivateThread = new AdapterInvokeThreadPtr(t);
-                t->start();
+                self->deactivateFuture = new std::future<void>();
+                *self->deactivateFuture = std::async(std::launch::async,
+                                                   [&self]
+                                                   {
+                                                       (*self->adapter)->waitForDeactivate();
+                                                   });
             }
 
-            while(!self->deactivated)
             {
-                bool done;
-                {
-                    AllowThreads allowThreads; // Release Python's global interpreter lock during blocking calls.
-                    done = (*self->deactivateMonitor).timedWait(IceUtil::Time::milliSeconds(timeout));
-                }
-
-                if(!done)
+                AllowThreads allowThreads; // Release Python's global interpreter lock during blocking calls.
+                if(self->deactivateFuture->wait_for(std::chrono::milliseconds(timeout)) == std::future_status::timeout)
                 {
                     PyRETURN_FALSE;
                 }
             }
-        }
 
-        assert(self->deactivated);
+            self->deactivated = true;
+            try
+            {
+                self->deactivateFuture->get();
+            }
+            catch(const Ice::Exception&)
+            {
+                *self->deactivateException = current_exception();
+            }
+        }
 
-        Ice::Exception* ex = (*self->deactivateThread)->getException();
-        if(ex)
+        assert(self->shutdown);
+        if(*self->deactivateException)
         {
-            setPythonException(*ex);
+            try
+            {
+                std::rethrow_exception(*self->deactivateException);
+            }
+            catch(const Ice::Exception& ex)
+            {
+                setPythonException(ex);
+            }
             return 0;
         }
     }
```

Comment:
should we catch std::exception here?

Reply:
Do, we only ever set an Ice::Exception in the exception_ptr